### PR TITLE
fix: JetStream context KV store/watch fix

### PR DESCRIPTION
**Pull request by chromevoid**
fix: #453 

# Test
## Settings
isbsvc
```yaml
#
# This manifest is intended for demonstration purpose, it's not suitable for production.
# Check https://numaflow.numaproj.io/user-guide/inter-step-buffer-service/ to figure out reliable configuration for production.
#
apiVersion: numaflow.numaproj.io/v1alpha1
kind: InterStepBufferService
metadata:
  name: default
spec:
  jetstream:
    version: latest # Do NOT use "latest" in real deployment, check "numaflow-controller-config" ConfigMap to get available versions.
    persistence:
      volumeSize: 1Gi
```
pipeline
```yaml
apiVersion: numaflow.numaproj.io/v1alpha1
kind: Pipeline
metadata:
  name: simple-pipeline
spec:
  vertices:
    - name: in
      source:
        # A self data generating source
        generator:
          rpu: 5
          duration: 1s
    - name: cat
      containerTemplate:
        env:
          - name: NUMAFLOW_DEBUG
            value: "true"
      scale:
        min: 1 # Minimum replicas
      udf:
        builtin:
          name: cat # A built-in UDF which simply cats the message
    - name: out
      containerTemplate:
        env:
          - name: NUMAFLOW_DEBUG
            value: "true"
      scale:
        min: 1 # Minimum replicas
      sink:
        # A simple log printing sink
        log: {}
  edges:
    - from: in
      to: cat
    - from: cat
      to: out
```
Disabled [nats native auto reconnect](https://github.com/numaproj/numaflow/blob/main/pkg/shared/clients/nats/default_jetstream_client.go#L54-L68) in `default_jetstream_client.go`. 

## Main branch
Used main branch to reproduce the bug:
- Started the isbsvc and the pipeline.Screen Shot 2023-01-10 at 6 24 48 PM
- Ran `kubectl scale sts isbsvc-default-js --replicas=0` to trigger [our reconnect](https://github.com/numaproj/numaflow/blob/main/pkg/shared/clients/nats/in_cluster_jetstream_client.go#L85-L109).
- After a few minutes, ran `kubectl scale sts isbsvc-default-js --replicas=3` to bring back the isbsvc.
- The service stopped working because KV store/watcher lost the new JetStream context.Screen Shot 2023-01-10 at 7 07 10 PM

## PR
Same steps as above. Still working after the reconnection.

Screen Shot 2023-01-10 at 7 42 25 PM

logs (updated the wording a little bit after the test)
```sh
2023-01-11T04:11:44.207Z	INFO	numaflow.MapUDF-processor	nats/in_cluster_jetstream_client.go:101	Succeeded to reconnect to Nat JetStream server	{"vertex": "simple-pipeline-cat"}
2023-01-11T04:11:45.210Z	INFO	numaflow.MapUDF-processor	jetstream/kv_watch.go:132	Succeeded to recreate the watcher	{"vertex": "simple-pipeline-cat", "pipeline": "simple-pipeline", "kvBucketName": "simple-pipeline-default-simple-pipeline-in-cat_PROCESSORS"}
2023-01-11T04:11:45.340Z	INFO	numaflow.MapUDF-processor	nats/in_cluster_jetstream_client.go:101	Succeeded to reconnect to Nat JetStream server	{"vertex": "simple-pipeline-cat"}
2023-01-11T04:11:45.342Z	INFO	numaflow.MapUDF-processor	jetstream/kv_store.go:63	Succeeded to recreate JetStream KeyValue store	{"vertex": "simple-pipeline-cat", "pipeline": "simple-pipeline", "bucketName": "simple-pipeline-default-simple-pipeline-cat-out_OT"}
2023-01-11T04:11:45.645Z	INFO	numaflow.MapUDF-processor	nats/in_cluster_jetstream_client.go:101	Succeeded to reconnect to Nat JetStream server	{"vertex": "simple-pipeline-cat"}
2023-01-11T04:11:46.495Z	INFO	numaflow.MapUDF-processor	nats/in_cluster_jetstream_client.go:101	Succeeded to reconnect to Nat JetStream server	{"vertex": "simple-pipeline-cat"}
2023-01-11T04:11:46.650Z	INFO	numaflow.MapUDF-processor	jetstream/kv_watch.go:132	Succeeded to recreate the watcher	{"vertex": "simple-pipeline-cat", "pipeline": "simple-pipeline", "kvBucketName": "simple-pipeline-default-simple-pipeline-in-cat_OT"}
2023-01-11T04:11:49.311Z	INFO	numaflow.MapUDF-processor	jetstream/kv_store.go:63	Succeeded to recreate JetStream KeyValue store	{"vertex": "simple-pipeline-cat", "pipeline": "simple-pipeline", "bucketName": "simple-pipeline-default-simple-pipeline-cat-out_PROCESSORS"}
```

### CODE REVIEW

File: pkg/watermark/store/jetstream/kv_store.go [hunk 1]
```diff
@@ -45,21 +48,33 @@ var _ store.WatermarkKVStorer = (*jetStreamStore)(nil)
...
+			// re-bind to an existing KeyValue store
+			kv, err := j.js.KeyValue(bucketName)
+			// keep looping because the watermark won't work without the store
+			for err != nil {
+				j.log.Errorw("Failed to rebind to the JetStream KeyValue store ", zap.Error(err))
+				kv, err = j.js.KeyValue(bucketName)
+				time.Sleep(100 * time.Millisecond)
+			}
+			j.log.Infow("Succeeded to rebind to JetStream KeyValue store")
+			j.kv = kv
```
Comment (vigith) [inline pkg/watermark/store/jetstream/kv_store.go:63]: there should be an RWlock here.

File: pkg/watermark/store/jetstream/kv_store.go [hunk 2]
```diff
@@ -69,28 +84,26 @@ func NewKVJetStreamKVStore(ctx context.Context, pipelineName string, bucketName
...
 	return j, nil
 }
 
-// JSKVStoreOption is to pass in Jetstream options.
+// JSKVStoreOption is to pass in JetStream options.
 type JSKVStoreOption func(*jetStreamStore) error
 
 // GetAllKeys returns all the keys in the key-value store.
 func (kv *jetStreamStore) GetAllKeys(_ context.Context) ([]string, error) {
+	kv.kvLock.Lock()
```
Comment (vigith) [inline pkg/watermark/store/jetstream/kv_store.go:105]: these locks should only be RLock

File: pkg/watermark/store/jetstream/kv_store.go [hunk 3]
```diff
@@ -36,6 +38,7 @@ type jetStreamStore struct {
 	pipelineName string
 	conn         *jsclient.NatsConn
 	kv           nats.KeyValue
+	kvLock       sync.Mutex
```
Comment (vigith) [inline pkg/watermark/store/jetstream/kv_store.go:41]: use RWLock? The code path is read heavy.

**Review by whynowy — APPROVED** [review]
LGTM

**Review by vigith — APPROVED** [review]

**Pull request merged**